Adding Arm Ethos-U NPU support

kristofer-jonsson-arm opened the pull request:
MPS3 AN547

- Updating mps3_an547 board files adding DTB entries for Ethos-U.
- Adding DTS bindings for the Ethos-U DTB entry.

Modules

- Adding module files for building the Arm Ethos-U Core driver.
- Updating TFLU module file to build Ethos-U operator.

Samples

- Adding TFLU Ethos-U sample program.

Manifest

- Adding hal_ethos_u to west manifest.

Signed-off-by: Kristofer Jonsson <kristofer.jonsson@arm.com>
Signed-off-by: Fredrik Knutsson <fredrik.knutsson@arm.com>

microbuilder commented on west.yml:
```diff
@@ -72,6 +72,11 @@ manifest:
       groups:
         - hal
+    - name: hal_ethos_u
+      revision: 7aeec657f2c1bb8d782bfc79413935a445081cfa
```
Just a heads up that the typical way to review changes like this is to generate a PR with the changes in the module, and then reference that PR in the west.yml manifest here via:

```
revision: pull/54/head
```

Where `54` is the PR number.

The changes were already pushed directly to `hal_ethos_u` so it's no longer relevant in this case, but it's useful to go through the PR route in case something also needs to be changed in the module code.

mbolivar-nordic commented:
Can you please rebase?

microbuilder commented:
@kristofer-jonsson-arm I have to setup the AN547 FVP image to properly test this since QEMU doesn't have U55 support, but looking through the code quickly, is `tflu` used elsewhere that you've seen? We've been using `tflm` in our APIs at Linaro, and it probably makes sense to be consistent one way or the other since this will all end up in Zephyr.

I usually see TensorFlow Lite Micro referenced as **TFLM** in documentation, and I suspect the `m` might be more recognisable to people?

kristofer-jonsson-arm commented:
Sorry for the late response. I was on vacation last week.

I have rebased the patch and also renamed TFLu to TFLM.

Please note as Kevin points out that QEMU will not be able to run the provided unit test, because it is lacking a NPU. For this the Corstone-300 FVP needs to be used.

https://developer.arm.com/tools-and-software/open-source-software/arm-platforms-software/arm-ecosystem-fvps

microbuilder commented:
@kristofer-jonsson-arm I've asked @theotherjimmy to have a look at this for testing, if you can just monitor the CI output and fix anything that shows up? Since you're a new committer, I might need to fire up the CI to run after any changes, though. Just ping me if I miss an update. Thanks!

theotherjimmy commented:
it looks to me like a step in the process is rebasing onto main, which seems to fail. Perhaps some of the merges are non-trivial?

theotherjimmy commented:
Since much of the commit log was trivial fixes or renames, I have squashed them all into a single commit. The branch is here: https://github.com/zephyrproject-rtos/zephyr/compare/main...theotherjimmy:ethos-u and you're welcome to push that branch over this one.

mbolivar-nordic commented:
@kristofer-jonsson-arm could you take a look at the twister errors in the CI results?

kristofer-jonsson-arm commented:
I'll have a look at the twister results.

The failing tests/lib/cmsis_dsp/filtering/libraries.cmsis_dsp.filtering.fir.fpu tests case seems to fail also if I revert my changes. But it might also be me doing something wrong.

I'll fix the coding style error and upload a new patch. It is is just one line causing a regression.

kristofer-jonsson-arm committed:
modules: add an Arm Ethos-U HAL module

Adding module files for building the Arm Ethos-U Core driver.
Updating TFLU module file to build Ethos-U operator.

Adding hal_ethos_u to west manifest.

Signed-off-by: Kristofer Jonsson <kristofer.jonsson@arm.com>
Signed-off-by: Fredrik Knutsson <fredrik.knutsson@arm.com>

kristofer-jonsson-arm committed:
dts: add Arm Ethos-U bindings

Updating mps3_an547 board files with DTB entries for Ethos-U.

Adding DTS bindings for the Ethos-U DTB entry.

Signed-off-by: Kristofer Jonsson <kristofer.jonsson@arm.com>
Signed-off-by: Fredrik Knutsson <fredrik.knutsson@arm.com>

kristofer-jonsson-arm committed:
samples: Add a TFLU Ethos-U sample program

Add a sample program that demonstates how to run inferences on
Arm Ethos-U.

Signed-off-by: Kristofer Jonsson <kristofer.jonsson@arm.com>
Signed-off-by: Fredrik Knutsson <fredrik.knutsson@arm.com>
Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

microbuilder commented on samples/tflm_ethosu/src/inference_process.cpp:
```diff
@@ -0,0 +1,221 @@
+	return false;
+}
+
+} // namespace
```
As per the [coding style requirements](https://docs.zephyrproject.org/latest/contribute/guidelines.html#coding-style):

> Use C89-style single line comments, /*  */. The C99-style single line comment, //, is not allowed.

Same issue elsewhere in this code.

microbuilder commented on dts/bindings/arm/arm,ethos-u.yaml:
```diff
@@ -0,0 +1,39 @@
+
+include: base.yaml
+
+properties:
+    reg:
+      required: true
+
+    interrupts:
+        required: true
```
Inconsistent leading whitespace.

kristofer-jonsson-arm replied:
Fixed.

kristofer-jonsson-arm commented on dts/bindings/arm/arm,ethos-u.yaml:
```diff
@@ -0,0 +1,39 @@
+        required: true
+
+    secure_enable:
+        type: int
```
This should be a boolean. I'll fix.

kristofer-jonsson-arm commented on dts/bindings/arm/arm,ethos-u.yaml:
```diff
@@ -0,0 +1,39 @@
+    secure_enable:
+        type: int
+        required: true
+        description: secure_enable
```
Absolutely. I'll add a better description.

kristofer-jonsson-arm commented on modules/hal_ethos_u/init.c:
```diff
@@ -0,0 +1,156 @@
+	const struct ethosu_data *data = dev->data;
+	struct ethosu_driver *drv = (struct ethosu_driver *)&data->drv;
+	struct ethosu_driver_version version;
+	printk("Ethos-U DTS info. base_address=0x%x, inst=%u, secure_enable=%u, privilege_enable=%u, irq=%u, "
```
Done.

mbolivar-nordic commented on modules/hal_ethos_u/init.c:
```diff
@@ -0,0 +1,164 @@
+		.inst = n,									   \
+	};											   \
+												   \
+	DEVICE_DT_INST_DEFINE(n, ethosu_zephyr_init, NULL, &ethosu_data_##n, &ethosu_dts_info_##n, \
+			      POST_KERNEL, CONFIG_KERNEL_INIT_PRIORITY_DEFAULT, NULL);
+
+DT_INST_FOREACH_STATUS_OKAY(ETHOSU_DEVICE_INIT);
```
What is going on here? Why is this file instantiating devices with NULL API pointers?

kristofer-jonsson-arm replied:
Thanks for taking time to review this code. This is the team's first contribution to Zephyr, so we are thankful for all the help we can get. And please bear with us if there are concepts in Zephyr that we have missed.

I didn't write this particular code myself, but it might be that we didn't know about the APIs when we implemented the device initialization.

Looking at the APIs under `zephyr/include/zephyr/drivers`, at first glance no API seems to be a perfect match Ethos-U. Ethos-U is a very specific piece of hardware that takes a command stream and up to 8 base pointers to compute a kernel in a neural network. If we have to implement a driver API, then I suspect we will need to add a new API just for Ethos-U.

The current implementation is very much based on how the Tensorflow Lite for Microcontrollers (TFLM) framework is designed. A neural network consists of a graph of kernels. The TFLM framework iterates of the nodes in the network and applies a kernel to each node. To accelerate a neural network on Ethos-U, the network model is run through the Vela compiler that replaces software kernels with a custom Ethos-U kernel. The implementation of the custom kernel can be found here. 

https://github.com/zephyrproject-rtos/tflite-micro/blob/main/tensorflow/lite/micro/kernels/ethos_u/ethosu.cc

The purpose of `init.c` is to register and enable the IRQ service route, and to initialize Ethos-U NPU driver based on the device tree entries.

mbolivar-nordic replied:
> Looking at the APIs under zephyr/include/zephyr/drivers, at first glance no API seems to be a perfect match Ethos-U. Ethos-U is a very specific piece of hardware that takes a command stream and up to 8 base pointers to compute a kernel in a neural network. If we have to implement a driver API, then I suspect we will need to add a new API just for Ethos-U.

OK. Then this feels like a misc driver to me (see drivers/misc/grove_lcd_rgb and include/zephyr/drivers/misc/grove_lcd/grove_lcd.h). @gmarull would you agree?

kristofer-jonsson-arm replied:
The NPU driver code is placed in a separate module, because it is mirroring an parent repository. It is only the init code that is stored in the main Zephyr repository.
https://github.com/zephyrproject-rtos/hal_ethos_u

Do you mean that the code should be moved from `modules/hal_ethos_u/` to `drivers/misc/`?

mbolivar-nordic replied:
Yes, I mean that if it is a driver in the sense of a zephyr device driver that allocates `struct device` objects, as it is currently structured, it should be in drivers/misc using the same style of API as the grove drivers I mentioned earlier.

mbolivar-nordic commented on arch/Kconfig:
```diff
@@ -196,6 +197,25 @@ config SRAM_BASE_ADDRESS
 	  changing it via menuconfig or in configuration files.
 
+# Workaround for not being able to have commas in macro arguments
+DT_CHOSEN_Z_DRAM := zephyr,dram
```
Any new chosen option needs to be documented in the list at the bottom of this page: https://docs.zephyrproject.org/latest/build/dts/api/api.html#devicetree-chosen-nodes

I would like to see a strong justification for why you need this option at all. The commit logs in the PR in general are too short and lack justification given that they are adding new chosen nodes and touching arch/Kconfig.

kristofer-jonsson-arm replied:
The Corstone-300 reference platform has beside TCM also SRAM and DRAM. We would like to control which memory the TFLM model is and arena are mapped to. How do we best do this?

mbolivar-nordic replied:
@carlocaione can you advise here?

carlocaione replied:
not sure if this exactly fits what you are trying to do but you can leverage the `zephyr,memory-region` compatible to create the DRAM region and allocate the proper sections in there.

See for example: https://github.com/zephyrproject-rtos/zephyr/blob/0c053190b79dd1f121be6e71e2d4fb2e2251fecb/boards/arm/stm32f429i_disc1/stm32f429i_disc1.dts#L26-L31 

and then:

https://github.com/zephyrproject-rtos/zephyr/blob/0c053190b79dd1f121be6e71e2d4fb2e2251fecb/drivers/memc/memc_stm32_sdram.ld#L19-L29

mbolivar-nordic replied:
> I reverted the changes to arch/Kconfig and replaced that with a custom linker script in the TFML Ethos-U sample application, similar to the memc example above.

But the problematic `zephyr,dram` chosen node still remains and therefore my comments about it are not yet addressed.

kristofer-jonsson-arm replied:
I have removed the `zephyr,dram` from the `chosen` section.

mbolivar-nordic commented on arch/Kconfig:
```diff
@@ -196,6 +197,25 @@ config SRAM_BASE_ADDRESS
+# Workaround for not being able to have commas in macro arguments
+DT_CHOSEN_Z_DRAM := zephyr,dram
+
+config DRAM_SIZE
```
I don't think you need a Kconfig option since this is being installed in devicetree from the beginning. The legacy SRAM options predate devicetree in Zephyr and are largely around for backwards compat.

mbolivar-nordic commented on dts/bindings/arm/arm,ethos-u.yaml:
```diff
@@ -0,0 +1,39 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+description: Ethos-U
```
Could you please explain more here?

kristofer-jonsson-arm replied:
Description has been updated.

mbolivar-nordic commented on dts/bindings/arm/arm,ethos-u.yaml:
```diff
@@ -0,0 +1,39 @@
+#
+# Copyright (c) 2022 Arm Limited.
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+# Licensed under the Apache License, Version 2.0 (the License); you may
+# not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an AS IS BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
```
General comment about this PR: I don't understand why it's mixing SPDX lines with verbatim licenses. This seems redundant and unnecessary. I think only the SPDX lines are needed. Am I missing something?

kristofer-jonsson-arm replied:
Agree the SPDX line should be sufficient. I'll update the files.

mbolivar-nordic commented on dts/bindings/arm/arm,ethos-u.yaml:
```diff
@@ -0,0 +1,39 @@
+    interrupts:
+        required: true
+
+    secure_enable:
+        type: boolean
+        required: true
+        description: Configure NPU to operate in secure- or non-secure mode
+
+    privilege_enable:
```
Devicetree property names should be separated with dashes (`-`), not underscores (`_`).

mbolivar-nordic reviewed (changes requested):
This PR seems to be in a very early state from the perspective of its commit logs. Many details are not justified or explained. Please see individual comments for details.

mansnils commented on samples/tflm_ethosu/prj.conf:
```diff
@@ -0,0 +1,6 @@
+#application default configuration
+# include TFLM based on CMSIS NN optimization and ETHOSU acceleration
```
TFLM has not yet CMSIS-NN support.

kristofer-jonsson-arm replied:
CMSIS-NN support has now been added.

carlocaione commented on samples/tflm_ethosu/linker.ld:
```diff
@@ -0,0 +1,17 @@
+	SECTION_DATA_PROLOGUE(_DDR4_SECTION_NAME,,SUBALIGN(16))
+	{
+        . = ALIGN(16);
+        *(tflm_model tflm_arena tflm_input tflm_output)
```
Just a comment here (nothing to change).

The `zephyr,memory-region` mechanism is already creating a section for you. The only difference with respect to this file is that in the section we are placing
```
*(DDR4 DDR4.*)
```
instead of:
```
*(tflm_model tflm_arena tflm_input tflm_output)
```
of course this is fine if keeping the naming is important for you.

kristofer-jonsson-arm replied:
Think it could be an advantage to keep the label, if we want to port the sample application to another target that has a different memory layout.

kristofer-jonsson-arm commented on boards/arm/mps3_an547/mps3_an547.dts:
```diff
@@ -85,6 +87,22 @@
+			interrupts = <56 3>;
+			secure-enable;
+			privilege-enable;
+			label = "ethosu0";
```
I'll remove the label.

XenuIsWatching commented on drivers/misc/ethos_u/init.c:
```diff
@@ -0,0 +1,152 @@
+#include <sys/util.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_DECLARE(ethos_u, CONFIG_KERNEL_LOG_LEVEL);
```
Shouldn't this be `LOG_MODULE_REGISTER` as this is the only log module for ethos

XenuIsWatching commented on drivers/misc/ethos_u/init.c:
```diff
@@ -0,0 +1,152 @@
+
+struct ethosu_dts_info {
+	uint32_t base_addr;
+	uint32_t secure_enable;
```
With them defined as bools do these still need to be uint32_t?

I also see in the ethos module that they are also uint32_t

kristofer-jonsson-arm replied:
I have changed the type to bool.

The driver interface still takes `uint32_t` and because it the Ethos-U module is a copy of upstream I don't think we for maintenance reasons should change this.

gmarull commented on drivers/misc/ethos_u/CMakeLists.txt:
```diff
@@ -0,0 +1,6 @@
+#
+# SPDX-License-Identifier: Apache-2.0
+
+zephyr_sources(
+    init.c)
```
please fix formatting here, a single line is enough

kristofer-jonsson-arm replied:
Fixed.

gmarull commented on drivers/misc/ethos_u/init.c:
```diff
@@ -0,0 +1,152 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "ethosu_driver.h"
+#include <device.h>
+#include <devicetree.h>
+#include <init.h>
+#include <kernel.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <sys/printk.h>
+#include <sys/util.h>
```
please prefix with `zephyr/`

gmarull commented on drivers/misc/ethos_u/init.c:
```diff
@@ -0,0 +1,152 @@
+#define ETHOSU_DEVICE_INIT(n)									   \
+	static void ethosu_zephyr_irq_config_##n(void);						   \
+												   \
+	static struct ethosu_data ethosu_data_##n = { .irq_config =				   \
+							      &ethosu_zephyr_irq_config_##n };	   \
```
why is a const pointer stored in data and not config?

gmarull commented on drivers/misc/ethos_u/init.c:
```diff
@@ -0,0 +1,152 @@
+	}											   \
+												   \
+	static const struct ethosu_dts_info ethosu_dts_info_##n = {				   \
+		.base_addr = DT_INST_REG_ADDR(n),						   \
```
you're later casting to `void *`, why not store a `void *` instead?

gmarull commented on drivers/misc/ethos_u/init.c:
```diff
@@ -0,0 +1,152 @@
+{
+	const struct ethosu_dts_info *config = dev->config;
+	const struct ethosu_data *data = dev->data;
+	struct ethosu_driver *drv = (struct ethosu_driver *)&data->drv;
```
redundant cast here

kristofer-jonsson-arm replied:
I have removed const from `dev->data` and the redundant cast.

gmarull commented on modules/hal_ethos_u/CMakeLists.txt:
```diff
@@ -0,0 +1,27 @@
+#
+# SPDX-License-Identifier: Apache-2.0
+
+cmake_policy(SET CMP0079 NEW)
```
cc: @tejlmand

gmarull commented on samples/tflm_ethosu/linker.ld:
```diff
@@ -0,0 +1,17 @@
+GROUP_START(DDR4)
+
+	SECTION_DATA_PROLOGUE(_DDR4_SECTION_NAME,,SUBALIGN(16))
+	{
+        . = ALIGN(16);
+        *(tflm_model tflm_arena tflm_input tflm_output)
```
please fix alignment here, this is mixing tabs/spaces

kristofer-jonsson-arm replied:
Will fix.

gmarull commented on samples/tflm_ethosu/src/inference_process.cpp:
```diff
@@ -0,0 +1,209 @@
+
+#include "inference_process.hpp"
+
+#include "tensorflow/lite/micro/all_ops_resolver.h"
+#include "tensorflow/lite/micro/cortex_m_generic/debug_log_callback.h"
+#include "tensorflow/lite/micro/micro_error_reporter.h"
+#include "tensorflow/lite/micro/micro_interpreter.h"
+#include "tensorflow/lite/micro/micro_profiler.h"
+#include "tensorflow/lite/schema/schema_generated.h"
```
these don't seem like local headers, use `<>`

kristofer-jonsson-arm replied:
Will fix.

gmarull commented on samples/tflm_ethosu/src/inference_process.cpp:
```diff
@@ -0,0 +1,209 @@
+			for (unsigned int j = 0; j < output->bytes; ++j) {
+				if (output->data.uint8[j] !=
+				    static_cast<uint8_t *>(expected.data)[j]) {
+					printk("Expected output tensor data mismatch. index=%u, offset=%u, expected=%02x, network=%02x\n",
```
do we have any recommendations for c++ styleguide @stephanosio ?

mbolivar-nordic commented on drivers/misc/ethos_u/CMakeLists.txt:
```diff
@@ -0,0 +1,5 @@
+#
+# SPDX-License-Identifier: Apache-2.0
+
+zephyr_sources(init.c)
```
Did you consider using zephyr_library() and zephyr_library_sources() instead? It's not clear to me why this needs to use zephyr_sources(), and that's quite unusual for a driver: compare the output of running `git grep zephyr_sources drivers` with `git grep zephyr_library_sources` in the zephyr top level directory for examples.

kristofer-jonsson-arm replied:
No particular reason for using `zephyr_sources`. If a Zephyr library is recommended I'll use that.

mbolivar-nordic commented on dts/bindings/arm/arm,ethos-u.yaml:
```diff
@@ -0,0 +1,26 @@
+#
+# SPDX-License-Identifier: Apache-2.0
+
+description: Arm Ethos-U NPU device tree binding
```
Sorry for being vague. When I asked "Could you please explain more here?" in https://github.com/zephyrproject-rtos/zephyr/pull/42580#discussion_r890621441, I more meant information like:

 - what is Ethos-U?
 - links to more information
 - whether users should define their own such nodes when configuring their devicetree, or if they should expect their board to define it for them and just modify an existing definition

Adding "Arm" at the beginning and "device tree binding" at the end is redundant without really clarifying what this binding is for -- does that make sense?

kristofer-jonsson-arm replied:
I have fleshed out the description a bit, clarifying what Ethos-U is and that it is expected to be configured part of the board device tree file.

I am a bit reluctant to add links to the description, as links tend to change and break over time.

Please have a look in next patch that is pushed.

mbolivar-nordic commented on dts/bindings/arm/arm,ethos-u.yaml:
```diff
@@ -0,0 +1,26 @@
+
+    secure-enable:
+        type: boolean
+        required: true
```
Combining `required: true` with boolean doesn't really make sense. Are you sure?

kristofer-jonsson-arm replied:
I have removed required, as you are perfectly right that it does not make any sense at all.

mbolivar-nordic commented on dts/bindings/arm/arm,ethos-u.yaml:
```diff
@@ -0,0 +1,26 @@
+
+    privilege-enable:
+        type: boolean
+        required: true
```
Are you sure?

mbolivar-nordic reviewed (changes requested):
Thank you both for your updates and your patience. This looks a lot closer to being mergeable from my point of view. Here is my next round of review. Thanks for your patience and persistence.

XenuIsWatching commented:
@kristofer-jonsson-arm, Any update on when the comments will be addressed from @mbolivar-nordic?

kristofer-jonsson-arm commented:
New patch has been uploaded. Please have a look.

mbolivar-nordic commented on drivers/misc/ethos_u/init.c:
```diff
@@ -0,0 +1,153 @@
+void *ethosu_mutex_create(void)
+{
+	int status = 0;
+	struct k_mutex *mutex = k_malloc(sizeof(*mutex));
```
The error handling in this function is not correct. k_malloc() is what can fail, not k_mutex_init(). This is missing a null check.

kristofer-jonsson-arm replied:
Done.

mbolivar-nordic commented on dts/bindings/arm/arm,ethos-u.yaml:
```diff
@@ -0,0 +1,32 @@
+    as a 'TFLM custom operation'.
+    
+    This device tree entry defines board specific properties like the register
+    map address and interrupt line, and is therefor expected to be included
```
Thank you, this description is excellent.

Typo: therefore

kristofer-jonsson-arm replied:
Done.

mbolivar-nordic commented on dts/bindings/arm/arm,ethos-u.yaml:
```diff
@@ -0,0 +1,32 @@
+    map address and interrupt line, and is therefor expected to be included
+    into the device specific device tree file.
+
+compatible: "arm,ethos-u"
```
The pull request is breaking git bisectability. The commit which adds the binding needs to come before the commit which adds the driver using the binding. Please watch https://www.youtube.com/watch?v=mnVNLM7ynMQ for more details.

kristofer-jonsson-arm replied:
Done. I have switched the order of the patches.

mbolivar-nordic commented on samples/tflm_ethosu/README.md:
```diff
@@ -0,0 +1,33 @@
+# Arm(R) Ethos(TM)-U Tensorflow Lite for Microcontrollers test application
```
This needs to be converted to .rst so it shows up in https://docs.zephyrproject.org/latest/samples/index.html.

See https://docs.zephyrproject.org/latest/contribute/documentation/index.html for details on contributing documentation.

kristofer-jonsson-arm replied:
I have never used rst before, but I hope I managed to correctly translate markdown into ReStructuredText.

mbolivar-nordic replied:
> I have never used rst before, but I hope I managed to correctly translate markdown into ReStructuredText.

You can check for yourself. I have just approved the run of a full CI on your PR. When the "Documentation build" step in the below list of CI checks is complete, you can click on the "details" link for that check, and the "summary" link on the resulting page. That should give you a link to a full build of the HTML documentation for your PR. Then you can verify that it is as you expect. If you want to build the documentation locally, you can do so using the instructions in the link I gave above, although this does take some time.

Please let me know when you are satisfied that your documentation renders correctly as HTML or if you have any specific questions about problems you had with your .rst.

mbolivar-nordic reviewed (changes requested):
Final change requests from my side. Thanks again.

kristofer-jonsson-arm commented:
`python3 ./scripts/ci/test_plan.py` fails with `Not enough FLASH but is one of the integration platforms`. How can I debug this?

Seems like Python package `gitpython` is missing from the requirements files under `scripts`.

stephanosio commented:
> `python3 ./scripts/ci/test_plan.py` fails with `Not enough FLASH but is one of the integration platforms`. How can I debug this?
> 
> Seems like Python package `gitpython` is missing from the requirements files under `scripts`.

That issue has been fixed. Rerunning CI should resolve the issue.

microbuilder commented:
> > `python3 ./scripts/ci/test_plan.py` fails with `Not enough FLASH but is one of the integration platforms`. How can I debug this?
> > Seems like Python package `gitpython` is missing from the requirements files under `scripts`.
> 
> That issue has been fixed. Rerunning CI should resolve the issue.

I kicked CI off again, although there are still some compliance issues to sort out.

kristofer-jonsson-arm commented on samples/tflm_ethosu/src/models/keyword_spotting_cnn_small_int8/input.h:
```diff
@@ -0,0 +1,41 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+__aligned(4) __attribute__((section("tflm_input"))) uint8_t inputData[] = {
```
The binary data originate from here https://github.com/ARM-software/ML-zoo/tree/master/models/keyword_spotting/cnn_small/tflite_int8`.

microbuilder replied:
Fixed link: https://github.com/ARM-software/ML-zoo/tree/master/models/keyword_spotting/cnn_small/tflite_int8

kristofer-jonsson-arm commented:
> Merging can be performed automatically once the requested changes are addressed.

Are there any more action from my side? One question is about C++ style guide, and the other a set of comments I believe have been addressed.

microbuilder commented:
> > Merging can be performed automatically once the requested changes are addressed.
> 
> 
> 
> Are there any more action from my side? One question is about C++ style guide, and the other a set of comments I believe have been addressed.

You need two authorized people to approve, and for the DNM tags to be removed, and this can be merged. Thanks for all the work on it.

mbolivar-nordic commented:
I will drop my -1 as soon as the TSC decision on whether we can allow neural network model data in samples is in (assuming the TSC approves this). There is a vote about this ongoing, which closes on October 25.

mbolivar-nordic commented:
@kristofer-jonsson-arm following discussion at the TSC approving a motion to allow neural network training data as binary blobs within samples as a special exception to our blobs policy, we can merge this, but you must add a `CONFIG_TAINT_BLOBS=y` from the application's Kconfig before we do. Otherwise you should be all set.

kristofer-jonsson-arm commented:
I have added `CONFIG_TAINT_BLOBS=y` to the project config, but then `west build -b mps3_an547 zephyr/samples/tflm_ethosu` fails with:

`error: TAINT_BLOBS (defined at modules/Kconfig:4) is assigned in a configuration file, but is not
directly user-configurable (has no prompt). It gets its value indirectly from other symbols. See
http://docs.zephyrproject.org/latest/kconfig.html#CONFIG_TAINT_BLOBS and/or look up TAINT_BLOBS in
the menuconfig/guiconfig interface. The Application Development Primer, Setting Configuration
Values, and Kconfig - Tips and Best Practices sections of the manual might be helpful too.
`

Any advice how I can fix this?

mbolivar-nordic commented:
> error: TAINT_BLOBS (defined at modules/Kconfig:4) is assigned in a configuration file, but is not directly user-configurable (has no prompt)

@carlescufi any objections to making this option directly settable so we can use that here to set TAINT_BLOBS=y in the prj.conf?

kristofer-jonsson-arm commented:
Any progress on `TAINT_BLOBS`? Would be nice to finally be able to merge this patch.

carlescufi commented:
> Any progress on `TAINT_BLOBS`? Would be nice to finally be able to merge this patch.

Sorry for the delay, I will get on this now.

mbolivar-nordic commented:
> I will double-check with @tejlmand

It was just an idea on my side; if there is a design violation here then I'm of course happy to see another better solution

mbolivar-nordic commented:
E.g. maybe have a new user-settable kconfig bool that selects `TAINT_BLOBS`, if we don't want to make it possible to 'un-select' TAINT_BLOBS directly?

tejlmand commented:
@kristofer-jonsson-arm what you should do is actually to create a Kconfig in the sample folder, like `zephyr/samples/tflm_ethosu/Kconfig`.

And then in that file write:
```
config TFLM_ETHOSU_TAINT_BLOBS
        bool
        default y
        select TAINT_BLOBS

source "Kconfig.zephyr"
```
that will set `TAINT_BLOBS` to `y` without the need to give the setting a prompt.
The reason for using the construct with a `select TAINT_BLOBS` is because it allows for other parts of the build system to actually see why TAINT_BLOBS was selected.

Whereas the option of just defaulting `TAINT_BLOBS` to `y` in the Kconfig doesn't provide the same tracing in other tooling.
Thus it becomes harder to trace back why `TAINT_BLOBS` is enabled.
(a bit philosophical discussion when talking about a sample that needs to select this, but this is the main reason for this way of using `select TAINT_BLOBS` and not having it user configurable).

carlescufi commented:
@kristofer-jonsson-arm can you fix the remaining CI issues?

`Error: * scan/samples/tflm_ethosu/Kconfig missing copyright.`

And then:
https://github.com/zephyrproject-rtos/zephyr/actions/runs/3426603450/jobs/5708607196
cc @stephanosio FYI.

carlocaione commented on dts/bindings/arm/arm,ethos-u.yaml:
```diff
@@ -0,0 +1,32 @@
+#
+# SPDX-License-Identifier: Apache-2.0
+
+description:
+    The Arm Ethos-U is a micro NPU that enables neural networks to be hardware
```
This should be `description: |`

kristofer-jonsson-arm replied:
Fixed

carlocaione reviewed (dismissed):
LGTM besides the one comment

stephanosio commented:
> And then:
> https://github.com/zephyrproject-rtos/zephyr/actions/runs/3426603450/jobs/5708607196
> cc @stephanosio FYI.

This is (sort of) a regression from https://github.com/zephyrproject-rtos/zephyr/pull/52052. I will send a fix to disable these CMSIS-DSP tests if we decide to keep the patches to `mps2_an521_remote` as-is.

https://discord.com/channels/720317445772017664/1014241011989487716/1039827287610241096

microbuilder commented:
> > And then:
> > https://github.com/zephyrproject-rtos/zephyr/actions/runs/3426603450/jobs/5708607196
> > cc @stephanosio FYI.
> 
> This is (sort of) a regression from #52052. I will send a fix to disable these CMSIS-DSP tests if we decide to keep the patches to `mps2_an521_remote` as-is.
> 
> https://discord.com/channels/720317445772017664/1014241011989487716/1039827287610241096

That PR is setting up the flash layout of `mps2_an521_ns` and `mps2_an521_remote` to a worst-case scenario where they are both used at once, with no overlap. If you aren't using `mps2_an521_ns` at the same time as `_remote`, you can reclaim the code space that `_ns` was using in a new custom overlay for the test case, setting:

> (Was 0x38B000 and 468K)

```
code: memory@200000 {
	reg = <0x00200000 DT_SIZE_M(2)>;
};
```

And changing the YAML to:

```yaml
ram: 512
flash: 2048
```

carlescufi commented:
@kristofer-jonsson-arm ignore the DNM label being re-added, the bot does that because you  are adding a new module to the manifest.

carlescufi commented:
@kristofer-jonsson-arm you still have some compliance issues:
https://github.com/zephyrproject-rtos/zephyr/actions/runs/3427710689/jobs/5711053603

carlescufi commented:
@stephanosio will you let us know once you've fixed the CMSIS-DSP failures?

stephanosio commented:
> @stephanosio @microbuilder will you let us know once you've fixed the CMSIS-DSP failures?

The hotfix just got merged: https://github.com/zephyrproject-rtos/zephyr/pull/52104

carlescufi commented:
@kristofer-jonsson-arm I just rebased your branch FYI.

gmarull commented on boards/arm/mps3_an547/mps3_an547.dts:
```diff
@@ -132,6 +132,7 @@
 	/* DDR4 - 2G, alternates non-secure/secure every 256M */
 	ddr4: memory@60000000 {
 		device_type = "memory";
```
First commit seems to be doing more than adding a binding, please, split commits properly.

carlescufi replied:
@kristofer-jonsson-arm if you need help splitting commits let us know, we can help.

carlescufi replied:
Fixed

gmarull commented on boards/arm/mps3_an547/mps3_an547.dts:
```diff
@@ -85,6 +86,21 @@
 		};
 	};
 
+	ethosu {
```
Question: shouldn’t this be defined at soc dts files (not board)?

carlescufi replied:
Possibly. But I think this is OK for something as new as this MPU. There's also no other boards with that SoC.

carlescufi replied:
There is no SoC .dtsi for this SoC, so I'd rather leave it here to be honest.

mbolivar-nordic replied:
Yeah, we seem to have done this inlining of the "soc" into the board for all the MPS2 AN style boards.

microbuilder replied:
I guess the assumption was that the 'SoC' would never exist outside a single combination of FPGA board + ANxxx bitstream. This particular board family is definitely a bit of a weird niche solution that found a second life solving specific CI problems.

gmarull commented on drivers/misc/ethos_u/CMakeLists.txt:
```diff
@@ -0,0 +1,6 @@
+#
+# SPDX-License-Identifier: Apache-2.0
+
+zephyr_library()
```
Again: this commit is doing more things other than adding a module. Please split changes accordingly.

carlescufi replied:
Fixed

gmarull commented on modules/hal_ethos_u/Kconfig:
```diff
@@ -0,0 +1,47 @@
+	help
+		Name of the used Arm NPU
+
+config ARM_ETHOS_U_LOG_LEVEL
```
Why isn’t this using logging module template files?

carlescufi replied:
I checked and no other module does, so I'm thinking this is overkill in this case.

gmarull replied:
why is it an overkill to do:

```
module = ARM_ETHOS_U
module-str = arm-ethos-u
source "subsys/logging/Kconfig.template.log_config"
```

it's actually shorter.

carlescufi commented:
@kristofer-jonsson-arm do you need assistance to complete the review by addressing @gmarull's review comments so we can merge this? Thanks!

kristofer-jonsson-arm commented:
> @kristofer-jonsson-arm do you need assistance to complete the review by addressing @gmarull's review comments so we can merge this? Thanks!

We have entered our release period, so I am afraid I don't have much more time to spend on this patch.

If it is necessary to split the commit, then I would be grateful for any help. I guess the DTS entry could/should be moved to `mps3_an547-common.dtsi`. I am not familiar with the logging template, but we need some kind of mapping between log levels defined by the NPU driver and Zephyr log levels.

carlescufi commented:
> > @kristofer-jonsson-arm do you need assistance to complete the review by addressing @gmarull's review comments so we can merge this? Thanks!
> 
> We have entered our release period, so I am afraid I don't have much more time to spend on this patch.
> 
> If it is necessary to split the commit, then I would be grateful for any help. I guess the DTS entry could/should be moved to `mps3_an547-common.dtsi`. I am not familiar with the logging template, but we need some kind of mapping between log levels defined by the NPU driver and Zephyr log levels.

OK, I will try to help you out with this if you don't mind me pushing to your branch @kristofer-jonsson-arm. Is that OK?

kristofer-jonsson-arm commented:
> OK, I will try to help you out with this if you don't mind me pushing to your branch @kristofer-jonsson-arm. Is that OK?

Please feel free to split and force push the review.

kristofer-jonsson-arm committed:
boards: mps3_an547: Add entry for Ethos-U

Add a DTB entry for the Ethos-U NPU present in the board.

Signed-off-by: Kristofer Jonsson <kristofer.jonsson@arm.com>

kristofer-jonsson-arm committed:
drivers: misc: Add an Ethos-U NPU driver

Add a driver for the Arm Ethos-U NPU, including a Devicetree entry for
the mps3_an547 platform.

Signed-off-by: Kristofer Jonsson <kristofer.jonsson@arm.com>
Signed-off-by: Fredrik Knutsson <fredrik.knutsson@arm.com>

gmarull commented on drivers/misc/ethos_u/init.c:
```diff
@@ -0,0 +1,159 @@
+
+	mutex = k_malloc(sizeof(*mutex));
+	if (mutex == NULL) {
+		LOG_ERR("Failed allocate mutex\n");
```
nit: redundant `\n`

gmarull commented on drivers/misc/ethos_u/init.c:
```diff
@@ -0,0 +1,159 @@
+
+	status = k_mutex_lock((struct k_mutex *)mutex, K_FOREVER);
+	if (status != 0) {
+		LOG_ERR("Failed to lock mutex with error - %d\n", status);
```
same here, and others

gmarull commented on drivers/misc/ethos_u/init.c:
```diff
@@ -0,0 +1,159 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "ethosu_driver.h"
```
this is not a local include, please use `<>`

gmarull commented on drivers/misc/ethos_u/init.c:
```diff
@@ -0,0 +1,159 @@
+#include <zephyr/devicetree.h>
+#include <zephyr/init.h>
+#include <zephyr/kernel.h>
+#include <zephyr/sys/printk.h>
```
this file is not printking, please, remove.

gmarull commented on drivers/misc/ethos_u/init.c:
```diff
@@ -0,0 +1,159 @@
+#include <zephyr/init.h>
+#include <zephyr/kernel.h>
+#include <zephyr/sys/printk.h>
+#include <zephyr/sys/util.h>
```
`zephyr/irq.h` is missing from the include list.

gmarull commented on drivers/misc/ethos_u/init.c:
```diff
@@ -0,0 +1,159 @@
+#include <zephyr/kernel.h>
+#include <zephyr/sys/printk.h>
+#include <zephyr/sys/util.h>
+#include <stdio.h>
```
why is this needed?

gmarull commented on drivers/misc/ethos_u/init.c:
```diff
@@ -0,0 +1,159 @@
+												   \
+	static void ethosu_zephyr_irq_config_##n(void)						   \
+	{											   \
+		IRQ_DIRECT_CONNECT(DT_INST_IRQN(n),						   \
```
question here (not blocking): is there any reason to use DIRECT IRQs?

carlescufi replied:
Let's leave it like this for now.

carlescufi commented:
@kristofer-jonsson-arm I have had to make quite a few changes to clean up your sample:

- Moved it to `samples/modules/tflite-micro`
- Added a `sample.yaml`
- Cleaned up the `CMakeLists.txt`

Aside from multiple changes in the driver itself.

carlescufi commented:
@microbuilder @kristofer-jonsson-arm now that the sample is actually building in CI, there are errors that must be looked at by those with platform knwoledge:

```
INFO    - 215/634 mps3_an547                zephyr/samples/modules/tflite-micro/tflm_ethosu/sample.drivers.tflm_ethosu FAILED Timeout (qemu 59.967s)
INFO    - /__w/zephyr/zephyr/twister-out/mps3_an547/zephyr/samples/modules/tflite-micro/tflm_ethosu/sample.drivers.tflm_ethosu/handler.log
ERROR   - E: Failed to switch security state and privilege level (ethosu_device_u55_u65.c:244)

E: Failed to initialize Ethos-U device (ethosu_driver.c:418)

*** Booting Zephyr OS build zephyr-v3.2.0-1611-g47a6e23aab63 ***
sender 0: Sending inference. job=0x30002a68, name=keyword_spotting_cnn_small_int8
runner 0: Received inference job. job=0x30002a68
W: No NPU driver handle available.

sender 0: Sending inference. job=0x30002ab4, name=keyword_spotting_cnn_small_int8
sender 1: Sending inference. job=0x30003270, name=keyword_spotting_cnn_small_int8
sender 1: Sending inference. job=0x300032bc, name=keyword_spotting_cnn_small_int8
[00:00:00.000,000] <err> ethos_u: Failed to initialize NPU with ethosu_init().
```

microbuilder commented:
> @microbuilder @kristofer-jonsson-arm now that the sample is actually building in CI, there are errors that must be looked at by those with platform knwoledge:

I'll have someone here dig into it, thanks for the cleanup.

carlescufi commented:
> EDIT: QEMU doesn't support the NPU, so that's the most likely explanation. This has to be run in the FVP. I'll get someone on the team with some experience with TFLM to kick the tires in, though.

Oh, it doesn't? Then this should be as easy as setting this as "build only" for now. Let me do that then.

microbuilder commented:
> > EDIT: QEMU doesn't support the NPU, so that's the most likely explanation. This has to be run in the FVP. I'll get someone on the team with some experience with TFLM to kick the tires in, though.
> 
> Oh, it doesn't? Then this should be as easy as setting this as "build only" for now. Let me do that then.

I've filed an action internally to get this working with the FVP in CI, but that is likely a new PR entirely, which I'll follow up separately and enable this sample in CI when that's working.

carlescufi commented:
> > > EDIT: QEMU doesn't support the NPU, so that's the most likely explanation. This has to be run in the FVP. I'll get someone on the team with some experience with TFLM to kick the tires in, though.
> > 
> > 
> > Oh, it doesn't? Then this should be as easy as setting this as "build only" for now. Let me do that then.
> 
> I've filed an action internally (to Linaro) to get this working with the FVP in CI, but that is likely a new PR entirely, which I'll follow up separately and enable this sample in CI when that's working.

Great! The sample will be built in CI, it just won't be run for now.

carlescufi reviewed (approved):
This is now good enough to get in, but please @kristofer-jonsson-arm refactor the sample so that it is generic, and uses the Ethos-U if available as a follow-up PR.
